src/PostgREST/App.hs

```diff
@@ -72,13 +77,16 @@ postgrest conf dbStructure pool =
     time <- getPOSIXTime
     body <- strictRequestBody req
 
-    let handleReq = runWithClaims conf time (app dbStructure conf body) req
+    let schema = cs $ configSchema conf
+        apiRequest = userApiRequest schema req body
+        handleReq = runWithClaims conf time (app dbStructure conf apiRequest) req
+
     resp <- either pgErrResponse id <$> P.use pool
-      (HT.run handleReq HT.ReadCommitted HT.Write)
+      (HT.run handleReq HT.ReadCommitted (transactionMode $ iAction apiRequest))
     respond resp
 
-app :: DbStructure -> AppConfig -> RequestBody -> Request -> H.Transaction Response
-app dbStructure conf reqBody req =
+app :: DbStructure -> AppConfig -> ApiRequest -> Request -> H.Transaction Response
+app dbStructure conf apiRequest req =
```

Comment:
Feels redundant to pass in both `apiRequest` and `req`. The latter is used for its `rawQueryString` to create an alphabetized canonical url for the resource. How about calculating that canonical url in `userApiRequest` and including it as a property of the `ApiRequest` itself?